### PR TITLE
Deprecate TLS older than 1.2

### PR DESCRIPTION
**Pull request by fenek**
Due to known vulnerabilities in 1.0 and 1.1.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[6809.2](https://travis-ci.org/esl/MongooseIM/jobs/562069016) / Erlang 22.0 / internal_mnesia / 290d9e96e0c5dd4ca3456236a1d3f4a3c44bd02a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2377/6809/internal_mnesia.22.0)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2377/6809/internal_mnesia.22.0/big//index.html)

**Comment by mongoose-im** [comment]
[6887.1](https://travis-ci.org/esl/MongooseIM/jobs/563109691) / Erlang 22.0 / small_tests / f72bc2fddc74e6d35b079da9459184b8d89c8590
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2377/6887/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2377/6887/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-07-24_16.38.07/lib.mongooseim.logs/run.2019-07-24_16.38.07/suite.log.html)

**Comment by mongoose-im** [comment]
[6906.1](https://travis-ci.org/esl/MongooseIM/jobs/564096080) / Erlang 22.0 / small_tests / c885212923ee70455ca43253eb8514254e420f21
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2377/6906/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2377/6906/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-07-26_17.04.02/lib.mongooseim.logs/run.2019-07-26_17.04.02/suite.log.html)

**Comment by mongoose-im** [comment]
[6910.1](https://travis-ci.org/esl/MongooseIM/jobs/564894489) / Erlang 22.0 / small_tests / f47624202be78d364052596683ae1289920f95f2
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2377/6910/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2377/6910/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-07-29_09.50.12/lib.mongooseim.logs/run.2019-07-29_09.50.13/suite.log.html)

File: big_tests/tests/connect_SUITE.erl [hunk 1]
```diff
@@ -79,30 +80,33 @@ groups() ->
     ct_helper:repeat_all_until_all_ok(G).
 
 tls_groups()->
-    [{group, c2s_noproc},
-     {group, starttls},
+    [{group, starttls},
+     {group, c2s_noproc},
```
Comment (fenek) [inline big_tests/tests/connect_SUITE.erl]: Meaningless change. I was reorganising groups and these two ended up in different order but it doesn't change anything.

**Comment by mongoose-im** [comment]
[6917.1](https://travis-ci.org/esl/MongooseIM/jobs/565077171) / Erlang 22.0 / small_tests / f19ba52ae1a2132bd3f3148324aa96e02b62c09c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2377/6917/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2377/6917/small_tests.22.0/small/ct_run.mongooseim@localhost.2019-07-29_17.52.08/lib.mongooseim.logs/run.2019-07-29_17.52.08/suite.log.html)

File: big_tests/tests/connect_SUITE.erl [hunk 2]
```diff
@@ -635,7 +642,7 @@ openssl_client_can_use_cipher(Cipher, Port) ->
     PortStr = integer_to_list(Port),
     Cmd = "echo '' | openssl s_client -connect localhost:" ++ PortStr ++
           " -cipher " "\"" ++ Cipher ++ "\" 2>&1",
-    {done, ReturnCode, _Result} = erlsh:oneliner(Cmd),
+    {done, ReturnCode, Result} = erlsh:oneliner(Cmd),
```
Comment (telezynski) [inline big_tests/tests/connect_SUITE.erl]: Result is unused

File: big_tests/tests/connect_SUITE.erl [hunk 3]
```diff
@@ -234,25 +235,35 @@ deny_pre_xmpp_1_0_stream(Config) ->
...
+
+should_fail_with_tlsv1(Config) ->
+    should_fail_with(Config, tlsv1).
+
+should_fail_with_tlsv1_1(Config) ->
+    should_fail_with(Config, 'tlsv1.1').
+
+should_fail_with(Config, Protocol) ->
+    %% Connection process is spawned with a link so besides the crash itself,
+    %% we will receive exit signal. We don't want to terminate the test due to this.
```
Comment (goddammit) [inline big_tests/tests/connect_SUITE.erl]:
```suggestion
    %%   we will receive an exit signal. We don't want to terminate the test due to this.
```

File: doc/advanced-configuration/Listener-modules.md [hunk 4]
```diff
@@ -16,32 +16,40 @@ You only need to declare running `ejabberd_c2s`, to have the other 2 modules sta
...
+* `starttls_required` (default: disabled) - enforces StartTLS usage.
+* `tls` (default: disabled) - When this option is set, clients must initiate a TLS session immediately after connecting, before beginning the normal XML stream.
+* `tls_module` (atom, default: `fast_tls`) - Provides a TLS library to use. `fast_tls` uses OpenSSL-based NIFs, while `just_tls` uses Erlang TLS implementation provided by OTP. They are fully interchangeable, with some exceptions (`ejabberd_c2s` options supported by only one of them are explicitly described, e.g. `crlfiles`).
+* `certfile` (string, default: no certfile will be used) - Path to the X509 PEM file with a certificate and a private key (not protected by a password).
+  If the certificate is signed by an intermediate CA, you should specify here the whole CA chain by concatenating all public keys together and appending the private key after that.
+* `cafile` (string, default: no CA file will be used) - Path to the X509 PEM file with a CA chain that will be used to verify clients. Won't have any effect if `verify_peer` is not enabled.
+* `verify_peer` (default: disabled) - Enforces verification of a client certificate. Requires valid `cafile`.
```
Comment (goddammit) [inline doc/advanced-configuration/Listener-modules.md]:
```suggestion
* `verify_peer` (default: disabled) - Enforces verification of a client certificate. Requires a valid `cafile`.
```

File: doc/advanced-configuration/TLS-hardening.md [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+Before we explain the TLS hardening in MongooseIM, we need to describe the TLS libraries used in the project.
+These are "OTP TLS" and "Fast TLS".
+
+The former is provided by (as the name suggests) OTP as the `ssl` application.
+Large part of the logic is implemented in Erlang but it calls OpenSSL API for some operations anyway.
+
+The latter is a community-maintained driver, which is implemented as NIFs (native C code).
+It uses OpenSSL API for all operations.
+
+Most MongooseIM components use TLS library provided by OTP.
```
Comment (goddammit) [inline doc/advanced-configuration/TLS-hardening.md]:
```suggestion
Most MongooseIM components use the TLS library provided by OTP.
```

File: doc/advanced-configuration/TLS-hardening.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    * Per socket in Fast TLS via OpenSSL cipher string
+
+## Deprecations
+
+MongooseIM is configured to allow only TLS 1.2 or higher, due to known vulnerabilities in TLS 1.0 and 1.1.
+It is still possible to enable earlier versions, however it is strongly discouraged.
+
+## OTP TLS hardening
+
+Protocols list for OTP TLS is set via `protocol_version` environment variable.
```
Comment (goddammit) [inline doc/advanced-configuration/TLS-hardening.md]:
```suggestion
Protocol list for OTP TLS is set via the `protocol_version` environment variable.
```

File: doc/advanced-configuration/TLS-hardening.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+It's an Erlang runtime variable, so it is not configured in OS but rather in `app.config` file.
```
Comment (goddammit) [inline doc/advanced-configuration/TLS-hardening.md]:
```suggestion
It's an Erlang runtime variable, so it is not configured in the OS but rather in the`app.config` file.
```

File: doc/advanced-configuration/TLS-hardening.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+## Fast TLS hardening
+
+Fast TLS expects an OpenSSL cipher string as one of optional connection parameters.
+This string is configured individually for every module that uses it.
+By default, MongooseIM sets this option to `TLSv1.2:TLSv1.3` for each component.
+
+The list below enumerates all components that use Fast TLS and describes how to change this string.
+
+* `ejabberd_c2s` - main user session abstraction + XMPP over TCP listener
+    * Please consult respective section in [Listener modules](Listener-modules.md#c2s-ciphers).
```
Comment (goddammit) [inline doc/advanced-configuration/TLS-hardening.md]:
```suggestion
    * Please consult the respective section in [Listener modules](Listener-modules.md#c2s-ciphers).
```

File: doc/advanced-configuration/TLS-hardening.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+* `ejabberd_s2s_in` - incoming S2S connections (XMPP Federation)
+    * Please consult respective section in [Listener modules](Listener-modules.md#s2s-ciphers).
```
Comment (goddammit) [inline doc/advanced-configuration/TLS-hardening.md]:
```suggestion
    * Please consult the respective section in [Listener modules](Listener-modules.md#s2s-ciphers).
```

**Pull request merged**